crates/winch/src/frame/mod.rs

```diff
@@ -0,0 +1,134 @@
+use crate::abi::{align_to, local::LocalSlot, ty_size, ABIArg, ABISig, ABI};
+use anyhow::Result;
+use smallvec::SmallVec;
+use std::ops::RangeInclusive;
+use wasmparser::{FuncValidator, FunctionBody, ValidatorResources};
+use wasmtime_environ::WasmType;
+
+// TODO:
+// SpiderMonkey's implementation uses 16; during instrumentation
+// we should measure to verify if this is a good default.
+pub(crate) type Locals = SmallVec<[LocalSlot; 16]>;
+
+/// Function defined locals start and end in the frame
+pub(crate) struct DefinedLocalsRange(pub RangeInclusive<u32>);
+
+/// Frame handler abstraction
+pub(crate) struct Frame {
+    /// The size of the entire local area; the arguments plus the function defined locals
+    pub locals_size: u32,
+
+    /// The range in the frame corresponding to the defined locals range
+    pub defined_locals_range: DefinedLocalsRange,
+
+    /// The local slots for the current function
+    ///
+    /// Locals get calculated when allocating a frame and are readonly
+    /// through the function compilation lifetime
+    pub locals: Locals,
+}
+
+impl Frame {
+    /// Allocate a new Frame
+    pub fn new<A: ABI>(
+        sig: &ABISig,
+        function: &mut FunctionBody,
+        validator: &mut FuncValidator<ValidatorResources>,
+        abi: &A,
+    ) -> Result<Self> {
+        let (mut locals, defined_locals_start) = Self::compute_arg_slots(sig, abi)?;
+        let (defined_slots, defined_locals_end) =
+            Self::compute_defined_slots(function, validator, defined_locals_start)?;
+        locals.extend(defined_slots);
+        let locals_size = align_to(defined_locals_end, abi.stack_align().into());
+
+        Ok(Self {
+            locals,
+            locals_size,
+            defined_locals_range: DefinedLocalsRange(defined_locals_start..=defined_locals_end),
+        })
+    }
+
+    /// Get a local slot
+    pub fn get_local(&self, index: u32) -> Option<&LocalSlot> {
+        self.locals.get(index as usize)
+    }
+
+    fn compute_arg_slots<A: ABI>(sig: &ABISig, abi: &A) -> Result<(Locals, u32)> {
+        // Go over the function ABI-signature and
+        // calculate the stack slots
+        //
+        //  for each parameter p; when p
+        //
+        //  Stack =>
+        //      The slot offset is calculated from the ABIArg offset
+        //      relative the to the frame pointer (and its inclusions, e.g.
+        //      return address)
+        //
+        //  Register =>
+        //     The slot is calculated by accumulating into the `next_frame_size`
+        //     the size + alignment of the type that the register is holding
+        //
+        //  NOTE
+        //      SpiderMonkey's implementation doesn't append any sort of
```

Comment:
References to SpiderMonkey are fine here but for the benefit of readers later, links to the source (via searchfox) might be helpful -- so e.g. starting out with "This algorithm is similar to SpiderMonkey's frame layout logic in LINK... The main differences are: ..."